contracts/Comptroller.sol

```diff
@@ -1150,151 +1081,167 @@ contract Comptroller is ComptrollerV4Storage, ComptrollerInterface, ComptrollerE
-            CToken cToken = allMarkets[i];
-            uint newSpeed = totalUtility.mantissa > 0 ? mul_(compRate, div_(utilities[i], totalUtility)) : 0;
-            compSpeeds[address(cToken)] = newSpeed;
-            emit CompSpeedUpdated(cToken, newSpeed);
-        }
+    function setCompSpeedInternal(CToken cToken, uint compSpeed) internal {
+        // note that COMP speed could be set to 0 to halt liquidity rewards for a market
+        compSpeeds[address(cToken)] = compSpeed;
+        emit CompSpeedUpdated(cToken, compSpeed);
     }
 
     /**
      * @notice Accrue COMP to the market by updating the supply index
      * @param cToken The market whose supply index to update
      */
     function updateCompSupplyIndex(address cToken) internal {
-        CompMarketState storage supplyState = compSupplyState[cToken];
-        uint supplySpeed = compSpeeds[cToken];
-        uint blockNumber = getBlockNumber();
-        uint deltaBlocks = sub_(blockNumber, uint(supplyState.block));
-        if (deltaBlocks > 0 && supplySpeed > 0) {
-            uint supplyTokens = CToken(cToken).totalSupply();
-            uint compAccrued = mul_(deltaBlocks, supplySpeed);
-            Double memory ratio = supplyTokens > 0 ? fraction(compAccrued, supplyTokens) : Double({mantissa: 0});
-            Double memory index = add_(Double({mantissa: supplyState.index}), ratio);
-            compSupplyState[cToken] = CompMarketState({
-                index: safe224(index.mantissa, "new index exceeds 224 bits"),
-                block: safe32(blockNumber, "block number exceeds 32 bits")
-            });
-        } else if (deltaBlocks > 0) {
-            supplyState.block = safe32(blockNumber, "block number exceeds 32 bits");
-        }
+        updateCompMarketIndex(cToken, compSupplyState[cToken], compSupplyVestingState[cToken], true, Exp({mantissa: 0}));
     }
 
     /**
      * @notice Accrue COMP to the market by updating the borrow index
      * @param cToken The market whose borrow index to update
      */
     function updateCompBorrowIndex(address cToken, Exp memory marketBorrowIndex) internal {
-        CompMarketState storage borrowState = compBorrowState[cToken];
-        uint borrowSpeed = compSpeeds[cToken];
+        updateCompMarketIndex(cToken, compBorrowState[cToken], compBorrowVestingState[cToken], false, marketBorrowIndex);
+    }
+
+    /**
+     * @notice Accrue COMP to the market by updating the supply or borrow index
+     * @param cToken The market whose index to update
+     * @param marketState The market state whose index to update
+     * @param vestingState The market vesting state whose index to update
+     * @param isSupply True if this implements the supply update, false if the borrow update
+     */
+    function updateCompMarketIndex(address cToken, CompMarketState storage marketState, CompMarketState storage vestingState, bool isSupply, Exp memory marketBorrowIndex) internal {
+        uint compSpeed = compSpeeds[cToken];
```

Comment:
Why not keep the old var name?

Reply:
@jflatow completely agree w/ the general principle of keeping variable names, in this case since the borrow and supply functions were abstracted into a `updateCompMarketIndex` function I didn't think either the `borrowSpeed` or `supplySpeed` was appropriate (assuming you're asking about `compSpeed`)!

Reply:
Oh I see. This looks like a very elaborate change now that's really going to require quite some extensive auditing. You've put a ton of thought into it, and that's amazing. I believe Coburn's suggestion was to split into 2 PRs - one with Comptroller reductions but no real changes in functionality, and another with the new vesting feature. I think its still true that splitting it up like that would be easier to integrate / digest. The math changes look pretty good to me, but as is it seems like they will actually add to code size not reduce it? Do we know what the net impact on code size + gas costs of these are?

Reply:
**Gas**
Thanks for these questions - taking in turn:

You can see the overall gas implications in the above comment:
https://github.com/compound-finance/compound-protocol/pull/71#issuecomment-710124317
(they were calculated by repurposing some of the original gas measurement scripts in the repo)

**Code size**
In terms of the math, we didn’t measure changes on a statement level for changes that were already “planned” as per “Eliminate ErrorReporter/Exponential” in https://www.comp.xyz/t/a-leaner-meaner-comptroller/461. But overall, a lot of the code size reductions came from these changes. As far as I remember, the `mul_ScalarTruncateAddUInt` definitely reduced overall code size.
Ah, I guess what you may be saying is that you expected most of the gains to surface when we completely eliminate the Error Reporter? We were pleasantly surprised that even these incremental substitutions did reduce overall footprint to the extent required!

**Audit**
In terms of auditing, both the Comptroller size reductions we had to do and the vesting changes are being audited by Quantstamp :).

As far as I understand, in the Compound dev call #1 it was decided that Gauntlet will mainly focus on getting the vesting changes in (together with some reductions that were required to free up space) while leaving comprehensive Comptroller code reductions for a bigger patch.